### PR TITLE
Feature/refactor worker part1

### PR DESCRIPTION
**Pull request by dereklwood**
This PR introduces the top-level worker class in the shared provisioner codebase.  This class, and the necessary supporting work to bring everything up to current functionality, are the only changes.  I tried to minimize changes for this first PR.
- [x] `lib/provisioner/worker.rb` replaces the old script: `lib/provisioner/worker/provisioner.rb`.  
  - [x] Code blocks were largely copy/pasted and just wrapped in functions where possible.
  - [x] bin/worker executable and supporting CLI
  - [x] handling of shared config, logging
- [x] supporting changes to `pluginmanager`, and wrapped in module (diff doesn't show well)
- [x] small changes to master (`tenantmanager` and `workerlauncher`) to launch new worker)
- [x] include `Coopr::Logging` in base automator/provider classes instead of plugins relying on global `log.x` methods

Remaining immediate work
- [ ] move base automator/provider, and plugins to Coopr::Plugins module
- [ ] move remaining supporting classes (signalhandler, etc) to Coopr module
- [ ] address all introduced TODOs in followup PRs

### CODE REVIEW

File: lib/provisioner/worker.rb [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+    # Instantiate and run an instance of given plugin for a task
+    def _run_plugin(clazz, env, cwd, task)
+      clusterId = task['clusterId']
+      hostname = task['config']['hostname']
+      provider = task['config']['provider']['description']
+      imagetype = task['config']['imagetype']
+      hardware = task['config']['hardwaretype']
+      taskName = task['taskName'].downcase
+      log.info "Creating node #{hostname} on #{provider} for #{clusterId} using #{imagetype} on #{hardware}" if taskName == 'create'
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:145]: Is this updated with @duddun 's changes?

Comment (dereklwood) [inline lib/provisioner/worker.rb:145]: yes

File: lib/provisioner/worker.rb [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+      end
+    end
+
+    def delegate_task(task, pluginmanager)
+      providerName = nil # rubocop:disable UselessAssignment
+      automatorName = nil # rubocop:disable UselessAssignment
+      clazz = nil # rubocop:disable UselessAssignment
+      object = nil
+      result = nil
+      classes = nil
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:162]: Maybe disable/enable the test around the block, rather than repeating it?

File: lib/provisioner/worker.rb [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+      log.debug "Processing task with id #{task_id} ..."
+
+      taskName = task['taskName'].downcase
+      # depending on task, these may be nil
+      # automator take pecedence as presence indicates a 'software' task
+      providerName = task['config']['provider']['providertype'] rescue nil
+      automatorName = task['config']['service']['action']['type'] rescue nil
+
+      case taskName.downcase
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb]: This downcase is redundant, since you downcase on assignment above on 167

File: lib/provisioner/worker.rb [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+          # server must specify which bootstrap handlers need to run
+          log.debug "Task #{task_id} running specified bootstrap handlers: #{task['config']['automators']}"
+          task['config']['automators'].each do |automator|
+            clazz = Object.const_get(pluginmanager.getHandlerActionObjectForAutomator(automator))
+            cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'automatortypes', automator)
+            result = _run_plugin(clazz, @plugin_env, cwd, task)
+            combinedresult.merge!(result)
+          end
+        else
+          log.warn 'No automators specified to bootstrap'
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:195]: Should this be a failure? If there's no automators to be run, there's no reason to bootstrap. If we're hitting this case, something is wrong, anyway.

File: lib/provisioner/worker.rb [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        sigterm = SignalHandler.new('TERM')
+        sigterm.dont_interupt {
+          result = delegate_task(task, @pluginmanager) # TODO: we dont need to pass pluginmanager anymore
+        }
+      rescue => e
+        log.error "Caught exception when running task from file #{@file}"
+
+        result = {} if result.nil? == true
+        result['status'] = '1'
+        if e.class.name == 'CommandExecutionError'
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:224]: What is the reason for this conditional? It's not entirely clear. A small comment would help.

File: lib/provisioner/worker.rb [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+      loop {
+        result = nil
+        response = nil
+        task = nil
+        begin
+          response = Coopr::RestHelper.post "#{@config.get(PROVISIONER_SERVER_URI)}/v2/tasks/take", { 'provisionerId' => @provisioner_id, 'workerId' => myid, 'tenantId' => @tenant }.to_json
+        rescue => e
+          log.error "Caught exception connecting to coopr server #{@config.get(PROVISIONER_SERVER_URI)}/v2/tasks/take: #{e}"
+          sleep 10
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb]: Maybe a TODO here... config knob for this timeout

File: lib/provisioner/worker.rb [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+            sleep 1
+            next
+          else
+            log.error "Received error code #{response.code} from coopr server: #{response.to_str}"
+            sleep 10
+            next
+          end
+        rescue => e
+          log.error "Caught exception processing response from coopr server: #{e.inspect}"
+        end
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:273]: Would it make sense to convert this to a method?

File: lib/provisioner/worker.rb [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+        # While provisioning, don't allow the provisioner to terminate by disabling signal
+        sigterm = SignalHandler.new('TERM')
+        sigterm.dont_interupt {
+          begin
+            result = delegate_task(task, @pluginmanager) # TODO: we dont need to pass pluginmanager anymore
+
+            result = Hash.new if result.nil? == true
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb:281]: Is this `== true` redundant?

File: lib/provisioner/worker/pluginmanager.rb [hunk 9]
```diff
@@ -21,151 +21,160 @@
...
+        return !@load_errors.empty?
+      end
+
+      # scan plugins directory for json plugin definitions, load plugins
+      def scan_plugins
+        # Allow both the old and new directory layouts
+        # old: ./plugins/['providers']/[plugin-name]/*.json new: ./plugins/[plugin-name]/*.json
+        (Dir["#{File.expand_path(File.dirname(__FILE__))}/plugins/*/*/*.json"] +
+         Dir["#{File.expand_path(File.dirname(__FILE__))}/plugins/*/*.json"] # +
+        # Add this back once we figure out how to pass the work_dir to PluginManager
```
Comment (wolf31o2) [inline lib/provisioner/worker/pluginmanager.rb]: Put TODO here so we can find it later.

File: lib/provisioner/worker.rb [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'providertypes', providerName)
+        result = _run_plugin(clazz, @plugin_env, cwd, task)
+      when 'install', 'configure', 'initialize', 'start', 'stop', 'remove'
+        clazz = Object.const_get(pluginmanager.getHandlerActionObjectForAutomator(automatorName))
+        cwd = File.join(@config.get(PROVISIONER_WORK_DIR), @tenant, 'automatortypes', automatorName)
+        result = _run_plugin(clazz, @plugin_env, cwd, task)
+      when 'bootstrap'
+        combinedresult = {}
+        classes = []
+        if task['config'].key? 'automators' and !task['config']['automators'].empty?
```
Comment (wolf31o2) [inline lib/provisioner/worker.rb]: Use `.key?('automators')` for consistency.

**Comment by dereklwood** [comment]
Regarding `lib/provisioner/worker.rb`, the contents of the functions were purposely unchanged from the current script, to minimize changes.  The purpose of this PR was just to move from a script to a class.  Anything unrelated to the changes, and not critical, I prefer to take note and address in a subsequent PR

File: lib/provisioner/worker/pluginmanager.rb [hunk 11]
```diff
@@ -21,151 +21,160 @@
...
-          # check ancestor to determine plugin type
-          klass = Object.const_get(jsondata[automatortype]['classname'])
-          if klass.ancestors.include? Object.const_get('Automator')
-            raise "plugin \"#{p_name}\" attempting to load duplicate automator type \"#{automatortype}\"" if @automatormap.key?(automatortype)
-            @automatormap.merge!({automatortype => jsondata[automatortype]})
+      def register_plugintype(name, json_obj, uri)
+        begin
+          log.debug "registering provider/automator type: #{name}"
+          json = JSON.generate(json_obj)
+          resp = Coopr::RestHelper.put("#{uri}", json, :'Coopr-UserID' => "admin", :'Coopr-TenantID' => "superadmin")
```
Comment (wolf31o2) [inline lib/provisioner/worker/pluginmanager.rb:131]: Assign the user/tenant somewhere and use it here, versus these hard-coded values. It'll make changing them later easier.

**Comment by wolf31o2** [comment]
Sure. Let's keep the changes for another PR. The only things I would go ahead and include is the TODO items, since those won't hurt to add now. That being said... :+1: LGTM

**Pull request merged**